File: js/data_structure/d_linked_list.js
class DoublyLinkedList {
	constructor(values){
		if(arguments.length > 0) {
			this.length = values.length;
			this.head = new DoublyLinkedList.Node(values[0]);
			let current = this.head;
			for (let i = 1; i < this.length; i++) {
				current.next = new DoublyLinkedList.Node(values[i],current);
				current = current.next;
			}
			this.tail = current;
		}
		else{
			this.length = 0;
			this.head = null;
			this.tail = null;
		}
	}

	static Node = class {
		constructor(val=null,prev_=null,next_=null){
			this.value=val;
			this.prev = prev_;
			this.next=next_;
		}
	}

	push(value){
		if(!this.head){
			this.head = new DoublyLinkedList.Node(value);
			this.tail = this.head;
		}
		else{
			let p = new DoublyLinkedList.Node(value,this.tail);
			this.tail.next = p;
			this.tail = p;
		}
		this.length++;
	}

	get(idx){
		if (idx < 0 || idx> this.length){
			return undefined;
		}
		let current = this.head;
		if(idx <= (this.length >>> 1) ){
			let i = 0;
			while(i !== idx){
				current = current.next;
				i++;
			}
		}
		else{
			let i = this.length - 1;
			current = this.tail;
			while (i !== idx){
				current = current.prev;
				i--;
			}
		}

		return current;
	}

	pop(){
		if(!this.head){
			return undefined;
		}
		let tmp = this.tail;
		if(this.length === 1){
			this.head=this.tail=null;
		}
		else{
			this.tail = tmp.prev;
			this.tail.next = null;
		}
		this.length--;
		return tmp.value;
	}

	/**
	 *
	 * @param idx Number
	 * @param value mixed
	 * @returns
	 */
	insert(idx,value){
		if(idx > this.length || idx < 0)
			return undefined;
		else if(idx === 0){
			this.unshift(value);
		}
		else if(idx === this.length){
			this.push(value);
		}
		else{
			let afterNode = this.get(idx);
			let beforeNode = afterNode.prev;
			let newNode = new DoublyLinkedList.Node(value, beforeNode, afterNode);
			beforeNode.next = newNode;
			afterNode.prev = newNode;
			this.length++;
		}
	}

	set(idx,value){
		if( idx > 0 && idx < this.length){
			let p = this.get(idx);
			p.value = value;
		}
	}

	shift(){
		if(this.head){
			let tmp = this.head;
			this.head = tmp.next;
			this.length--;
			if(this.length === 0){
				this.tail = undefined;
			}
			return tmp;
		}
	}

	unshift(value){
		if(this.head){
			this.head = new DoublyLinkedList.Node(value,null,this.head);
		}
		else{
			this.head = new DoublyLinkedList.Node(value);
			this.tail = this.head;
		}
		this.length++;
	}

	remove(idx){
		if(idx < 0 || idx > this.length){
			return undefined;
		}
		else if(idx === this.length - 1){
			this.pop();
		}
		else if(idx === 0){
			this.shift();
		}
		else{
			let prev = this.get(idx-1);
			let remove = prev.next;
			prev.next = remove.next;
			this.length--;
		}
	}

	reverse(){
		let current = this.head
		this.head = this.tail
		this.tail = current

		let prev = null, next = null

		while (current !== null){
			prev = current.prev
			next = current.next
			current.prev = next
			current.next = prev
			current = next
		}
	}

	toString(){
		let current = this.head;
		let output = '['
		for(let i=0;i<this.length;i++){
			output += current.value.toString();
			if(i < this.length - 1){
				output += ', ';
			}
			current = current.next;
		}
		output += ']';
		return output;
	}
	display(){
		console.log(this.toString());
	}

}
function test(){
	const ll = new DoublyLinkedList([4,5]);
	ll.push(6);
	ll.insert(2,222);
	ll.push(7);
	ll.push(8);
	ll.push(9);
	ll.push(10);
	ll.reverse();
	ll.pop();
	ll.set(2,100);
	ll.insert(5,555);
	ll.insert(4,444);
	ll.shift();
	ll.unshift(6);
	ll.remove(5);
	const expected = '[6, 9, 100, 7, 444, 555, 222, 5]';
	if(ll.toString() === expected ){
		console.log("All's clear");
	}
	else{
		console.log(`We got ${ll.toString()} but we expected ${expected}.`)
	}
}
if (typeof module !== 'undefined' && require.main === module) {
	test();
}

File: lua/linked_list/d_linked_list.lua
--[[
By Macarthur Inbody
License LGPLv3 6/6/21

]]

--package.path = package.path .. ";../?.lua"
require 'util'

--[[
methods to write
pop()
push(value)
shift() get the value at the head and remove it
unshift() place value at the head
insert(idx,value) adds a value at a certain index
get(idx) gets the node at the referred to index
remove(idx) removes the node at the specified index
reverse() reverses the LinkedList in place
]]

DblLinkedList = Class{}

function DblLinkedList:init()
	self.Node = Class{
		init = function(self,value,prev,next)
			self.value = value
			self.prev = prev
			self.next = next
		end
	}
	self._head = nil
	self._tail = nil
	self._length = 0
end

-- remove the head node
function DblLinkedList:shift()
	if self._head then
		local tmp = self._head
		self._head = tmp.next
		self._head.prev = nil
		self._length = self._length - 1
		if self._length == 0 then
			self._tail = nil
		end
		return tmp
	else
		return nil
	end
end

function DblLinkedList:unshift(value)
	if self._head then
		self._head = self.Node(value,nil,self._head)
	else
		self._head = self.Node(value)
		self._tail = self._head
	end
end

function DblLinkedList:push(value)
	if self._head == nil then
		self._head = self.Node(value)
		self._tail = self._head
	else
		local p  = self.Node(value,self._tail)
		self._tail.next = p
		self._tail = self._tail.next
	end
	self._length = self._length + 1
end

function DblLinkedList:pop(idx)
	local popped
	if self._head == nil then
		return nil
	end
	if idx and (idx > 1 or idx < self._length) then
		popped = self:get(idx)
	else
		popped = self._tail
		self._tail = self._tail.prev
		self._tail.next = nil
	end
	self._length = self._length - 1
	return popped
end

function DblLinkedList:get(idx)
	if idx < 1 or idx > self._length then
		return nil
	end
	local cur
	if idx > (self._length / 2) then
		cur = self._tail
		for i=1,idx-1 do
			cur = cur.prev
		end
	else
		cur = self._head
		for i=1,idx-1 do
			cur = cur.next
		end
	end
	return cur
end

function DblLinkedList:stringify()
	local os = '['
	local cur = self._head
	while cur.next ~= nil do
		os = os .. tostring(cur.value) .. ","
		cur = cur.next
	end
	return os .. tostring(cur.value) .. "]"
end

function DblLinkedList:insert(idx,value)
	if 1 > idx or idx > self._length then
		return nil
	end
	local cur = self:get(idx)
	local old_prev = cur.prev
	local new_node = self.Node(value,old_prev, cur)
	old_prev.next = new_node
	cur.prev = new_node
	self._length = self._length + 1
end

-- change the value of the node at the given index to the value given
function DblLinkedList:set(idx,value)
	if idx < 1 or idx > self._length then
		return nil
	end
	local Node = self:get(idx)
	Node.value = value
end

function DblLinkedList:reverse()
	local prev = nil
	local cur = self._head
	local next = nil
	self._head = self._tail
	self._tail = cur
	while cur do
		next = cur.next
		cur.next = prev
		cur.prev = next
		prev = cur
		cur = next
	end
end

function DblLinkedList:len()
	return self._length
end

if not pcall(debug.getlocal,4,1) then
	ll = DblLinkedList()
	ll:push(1)
	ll:push(2)
	ll:push(3)
	ll:push(4)
	ll:push(5)
	ll:push(6)
	print(ll:stringify())
	ll:pop()
	ll:pop()
	print(ll:stringify())
	ll:shift()
	ll:unshift(11)
	print(ll:stringify())
	ll:pop(2)
	print(ll:stringify())
	-- nil test
	print(ll:get(0))
	ll:insert(2,-11)
	print(ll:stringify())
	ll:reverse()
	print(ll:stringify())
end
